python/lsst/daf/butler/gen2convert/writer.py

```diff
@@ -0,0 +1,316 @@
+        self.repos[gen2repo.root] = converted
+        return converted
+
+    def run(self, registry, datastore):
+        """Main driver for ConversionWriter.
+
+        Runs all steps to create a Gen3 Repo, aside from Camera registration
+        (we merely check that the needed Cameras, Sensors, and PhysicalFilters
+        have already been registered).
+        """
+        self.checkCameras(registry)
+        self.insertSkyMaps(registry)
+        self.insertObservations(registry)
+        self.insertDatasetTypes(registry)
+        self.insertDatasets(registry, datastore)
+        # TODO: associate parent repo datasets with child repo Collections
+
+    def checkCameras(self, registry):
+        """Check that all necessary Cameras are already present in the
+        Registry.
+        """
+        log = Log.getLogger("lsst.daf.butler.gen2convert")
+        cameras = set()
+        for repo in self.repos.values():
+            cameras.add(self.config["mappers"][repo.gen2.MapperClass.__name__]["camera"])
+        for camera in cameras:
+            log.debug("Looking for preexisting Camera '%s'.", camera)
+            if registry.findDataUnitEntry("Camera", {"camera": camera}) is None:
+                raise LookupError(
+                    "Camera '{}' has not been registered with the given Registry.".format(camera)
+                )
+
+    def insertSkyMaps(self, registry):
+        """Add all necessary SkyMap DataUnits (and associated Tracts and
+        Patches) to the Registry.
+        """
+        log = Log.getLogger("lsst.daf.butler.gen2convert")
+        for sha1, skyMap in self.skyMaps.items():
+            skyMapName = self.skyMapNames.get(sha1, None)
+            try:
+                existing, = registry.query("SELECT skymap FROM SkyMap WHERE sha1=:sha1",
+                                           sha1=sha1)
+                if skyMapName is None:
+                    skyMapName = existing["skymap"]
+                    self.skyMapNames[sha1] = skyMapName
+                    log.debug("Using preexisting SkyMap '%s' with sha1=%s", skyMapName, sha1.hex())
+                if skyMapName != existing["skymap"]:
+                    raise ValueError(
+                        ("SkyMap with new name={} and sha1={} already exists in the Registry "
+                         "with name={}".format(skyMapName, sha1.hex(), existing["skymap"]))
+                    )
+                continue
+            except ValueError:
+                # No SkyMap with this sha1 exists, so we need to insert it.
+                pass
+            if skyMapName is None:
+                raise LookupError(
+                    ("SkyMap with sha1={} has no name "
+                     "and does not already exist in the Registry.").format(sha1.hex())
+                )
+            log.info("Inserting SkyMap '%s' with sha1=%s", skyMapName, sha1.hex())
+            skyMap.register(skyMapName, registry)
```

Comment:
Do we want the same skymap to have different names? Otherwise I don't really get why the call can't be `skyMap.register(registry)`.

Reply:
Or directly `addDataUnitEntry(...)`.